demosplan/DemosPlanCoreBundle/Logic/OwnsProcedureConditionFactory.php

```diff
@@ -13,120 +13,154 @@
-            $procedurePlanningOffices,
-            ['orga', 'id']
-        );
+            return $this->conditionFactory->propertyHasStringAsMember($organisationId, ['planningOffices']);
+        }
+
+        $procedure = $this->userOrProcedure;
+        $procedurePlanningOffices = $procedure->getPlanningOfficesIds();
+
+        return $this->conditionFactory->propertyHasAnyOfValues($procedurePlanningOffices, ['orga', 'id']);
     }
 
     /**
-     * If {@link GlobalConfigInterface::hasProcedureUserRestrictedAccess} is set to `false`
+     * If {@link GlobalConfigInterface::hasProcedureUserRestrictedAccess} is set to `false`,
      * then the user must be in the organisation that created the procedure.
      *
-     * If {@link GlobalConfigInterface::hasProcedureUserRestrictedAccess} is set to `true`
-     * then the user must either be authorized {@link OwnsProcedureConditionFactory::isAuthorizedViaPlanningAgency()
-     * via their planning agency} or manually **regardless of their role**.
+     * If {@link GlobalConfigInterface::hasProcedureUserRestrictedAccess} is set to `true`,
+     * then the user must be authorized manually for the procedure.
+     *
+     * The returned condition will not apply role checks by itself. Use in conjunction with
+     * {@link self::hasProcedureAccessingRole}.
      *
      * @return FunctionInterface<bool>
      */
     public function isAuthorizedViaOrgaOrManually(): FunctionInterface
     {
-        $orgaOwnsProcedure = $this->conditionFactory->propertyHasValue(
-            $this->procedure->getOrgaId(),
-            ['orga', 'id']
-        );
-
-        // T8427: allow access by manually configured users, overwriting the organisation-based access
-        if ($this->globalConfig->hasProcedureUserRestrictedAccess()) {
-            $planningAgencyIsAuthorized = $this->isAuthorizedViaPlanningAgency();
-            $userIsAuthorized = $this->conditionFactory->propertyHasAnyOfValues(
-                $this->procedure->getAuthorizedUserIds(),
-                ['id']
-            );
-
-            $orgaOwnsProcedure = $this->conditionFactory->anyConditionApplies(
-                $userIsAuthorized,
-                $planningAgencyIsAuthorized
-            );
-        }
-
-        return $orgaOwnsProcedure;
+        // T8427: allow access by manually configured users if the config is set to `true`,
+        // overwriting the organisation-based access
+        return $this->globalConfig->hasProcedureUserRestrictedAccess()
+            ? $this->userIsExplicitlyAuthorized()
+            : $this->userOwnsProcedureViaOrgaOfUserThatCreatedTheProcedure();
     }
 
     /**
-     * Returns a condition to match users having the roles to theoretically own a procedure.
+     * Returns a condition to match users having the roles in the given customer to theoretically own a procedure.
      *
-     * @return FunctionInterface<bool>
+     * @return list<FunctionInterface<bool>>
```

Comment:
is this correct?

Reply:
`list` is a type introduced by phpstan and as far as I know accepted by PHPStorm. It denotes an array of values whose keys are integers starting with `0` that increase without holes.

`FunctionInterface<bool>` is "good enough". There are inconsistencies when to use `FunctionInterface` vs `ClauseFunctionInterface` (the latter extending the former). There is a clear solution for that, but it requires some refactoring which execution needs to be planned and approved first.

Reply:
`list` is a type introduced by phpstan and as far as I know accepted by PHPStorm. It denotes an array of values whose keys are integers starting with `0` that increase without holes.

`FunctionInterface<bool>` is "good enough". There are inconsistencies when to use `FunctionInterface` vs `ClauseFunctionInterface` (the latter extending the former). There is a clear solution for that, but it requires some refactoring which execution needs to be planned and approved first.